flux-middle/src/rty/projections.rs

```diff
@@ -48,22 +87,254 @@ fn without_constrs<T: TypeFoldable>(t: &T) -> T {
+        GenericArg::Lifetime(_) => todo!(),
+    }
+}
+fn into_rustc_bty<'tcx>(tcx: &TyCtxt<'tcx>, bty: &BaseTy) -> rustc_middle::ty::Ty<'tcx> {
+    match bty {
+        BaseTy::Int(i) => rustc_middle::ty::Ty::new_int(*tcx, *i), // rustc_middle::ty::Ty::mk_int(*tcx, int_ty),
+        BaseTy::Uint(i) => rustc_middle::ty::Ty::new_uint(*tcx, *i),
+        BaseTy::Param(pty) => pty.to_ty(*tcx),
+        BaseTy::Slice(ty) => rustc_middle::ty::Ty::new_slice(*tcx, into_rustc_ty(tcx, ty)),
+        BaseTy::Bool => todo!(),
+        BaseTy::Str => rustc_middle::ty::Ty::new_static_str(*tcx),
+        BaseTy::Char => todo!(),
+        BaseTy::Adt(adt_def, substs) => {
+            let did = adt_def.did();
+            let adt_def = tcx.adt_def(did);
+            let substs = substs.iter().map(|arg| into_rustc_generic_arg(tcx, arg));
+            let substs = tcx.mk_args_from_iter(substs);
+            rustc_middle::ty::Ty::new_adt(*tcx, adt_def, substs)
+        }
+        BaseTy::Float(f) => rustc_middle::ty::Ty::new_float(*tcx, *f),
+        BaseTy::RawPtr(_, _) => todo!(),
+        BaseTy::Ref(_, _, _) => todo!(),
+        BaseTy::Tuple(_) => todo!(),
+        BaseTy::Array(_, _) => todo!(),
+        BaseTy::Never => todo!(),
+        BaseTy::Closure(_, _) => todo!(),
+        BaseTy::Alias(_, _) => todo!(),
+    }
+}
+
+fn into_rustc_ty<'tcx>(tcx: &TyCtxt<'tcx>, ty: &Ty) -> rustc_middle::ty::Ty<'tcx> {
+    match ty.kind() {
+        TyKind::Indexed(bty, _) => into_rustc_bty(tcx, bty),
+        TyKind::Exists(ty) => into_rustc_ty(tcx, &ty.clone().skip_binder()),
+        TyKind::Constr(_, ty) => into_rustc_ty(tcx, ty),
+        TyKind::Param(pty) => pty.to_ty(*tcx),
+        TyKind::Uninit => todo!(),
+        TyKind::Ptr(_, _) => todo!(),
+        TyKind::Discr(_, _) => todo!(),
+        TyKind::Downcast(_, _, _, _, _) => todo!(),
+        TyKind::Blocked(_) => todo!(),
+    }
+}
+
+#[derive(Debug)]
+struct TVarSubst {
+    map: FxHashMap<ParamTy, Ty>,
+}
+
+impl TVarSubst {
+    fn mk_subst(src: &rustc_middle::ty::Ty, dst: &Ty) -> Vec<GenericArg> {
+        let mut subst = TVarSubst { map: FxHashMap::default() };
+        subst.infer_from_ty(src, dst);
+        subst
+            .map
+            .iter()
+            .sorted_by_key(|(pty, _)| pty.index)
+            .map(|(_, ty)| GenericArg::Ty(ty.clone()))
+            .collect()
+    }
+
+    fn insert(&mut self, pty: &ParamTy, ty: &Ty) {
+        match self.map.insert(*pty, ty.clone()) {
+            None => (),
+            Some(_) => bug!("duplicate insert"),
+        }
+    }
+
+    fn infer_from_arg(&mut self, src: &rustc_middle::ty::GenericArg, dst: &GenericArg) {
+        if let GenericArg::Ty(dst) = dst {
+            self.infer_from_ty(&src.as_type().unwrap(), dst)
+        }
+    }
+
+    fn infer_from_ty(&mut self, src: &rustc_middle::ty::Ty, dst: &Ty) {
+        use rustc_middle::ty;
+        match src.kind().clone() {
+            ty::TyKind::Param(pty) => self.insert(&pty, dst),
```

Comment:
I think we need to check for escaping vars here when we are actually matching it with a type variable.